src/Microsoft.Azure.ServiceBus.MessageId/MessageIdPlugin.cs

```diff
@@ -29,7 +29,23 @@ public class MessageIdPlugin : ServiceBusPlugin
         /// <param name="messageIdGenerator">Message ID generator to use.</param>
         public MessageIdPlugin(Func<Message, string> messageIdGenerator)
         {
-            this.messageIdGenerator = messageIdGenerator;
+            Guard.AgainstNull(nameof(messageIdGenerator), messageIdGenerator);
+            this.messageIdGenerator = SafeMessageIdGenerator(messageIdGenerator);
+        }
+
+        private Func<Message, string> SafeMessageIdGenerator(Func<Message, string> originalMessageIdGenerator)
+        {
+            return message =>
+            {
+                try
+                {
+                    return originalMessageIdGenerator(message);
+                }
+                catch (Exception exception)
+                {
+                    throw new Exception("An exception occurred when executing message ID generator Func", exception);
+                }
+            };
         }
```

Comment:
Potentially 2 reasons @nemakam:
1. If you wrap `messageIdGenerator` with `try-catch`, you have to do it everywhere you use it. In this specific instance it's only one location. But if used more than once, more than one location would require that wrapping.
2. To provide better feedback when a failure is taking place. 

#2 is probably the most important. When an exception is thrown with `try-catch` wrapping the original invocation, stacktrace will point at the method that was using `messageIdGenerator`. When you have a safe method, the stack trace will explicitly show that on top, helping to diagnose what went wrong.

![image](https://user-images.githubusercontent.com/1309622/27893889-f63103ba-61c5-11e7-97d7-52ae49f18394.png)

which is more user guiding than the alternative

![image](https://user-images.githubusercontent.com/1309622/27893904-1a354906-61c6-11e7-9948-e459ad9b04f0.png)